gapis/api/vulkan/frame_loop.go

```diff
@@ -126,129 +139,237 @@ func newFrameLoop(ctx context.Context, c *capture.GraphicsCapture, numInitialCmd
-		log.D(ctx, "Loop: start loop at frame %v, id %v, cmd %v.", f.frameNum, id, cmd)
-		f.detectChangedResource(ctx, out.State())
+	if lastObservedCommand != api.CmdNoID && lastObservedCommand > api.CmdID.Real(cmdId) {
+		log.F(ctx, true, "FrameLoop: expected next observed command ID to be >= last observed command ID")
+	}
+
+	// Walk the frame count forwards if we just hit the end of one.
+	if _, ok := cmd.(*VkQueuePresentKHR); ok {
+		f.frameNum++
+	}
 
-		st := GetState(out.State())
-		sb := st.newStateBuilder(ctx, newTransformerOutput(out))
-		defer sb.ta.Dispose()
+	// Are we before the loop or just at the start of it?
+	if lastObservedCommand == api.CmdNoID {
+
+		// This is the start of the loop.
+		if api.CmdID.Real(cmdId) == f.loopStartIdx {
+
+			log.D(ctx, "FrameLoop: start loop at frame %v, cmdId %v, cmd %v.", f.frameNum, cmdId, cmd)
+
+			f.capturedLoopCmds = append(f.capturedLoopCmds, cmd)
+			f.capturedLoopCmdIds = append(f.capturedLoopCmdIds, cmdId)
 
-		if err := f.backupChangedResources(ctx, sb); err != nil {
-			log.E(ctx, "Failed to backup changed resources: %v", err)
+			return
+
+		} else {
+			// The current command is before the loop begins and needs no special treatment. Just pass-through.
+			log.D(ctx, "FrameLoop: before loop at frame %v, cmdId %v, cmd %v.", f.frameNum, cmdId, cmd)
+			out.MutateAndWrite(ctx, cmdId, cmd)
 			return
 		}
-		// Add jump label
-		sb.write(sb.cb.Custom(func(ctx context.Context, s *api.GlobalState, b *builder.Builder) error {
-			f.loopCountPtr = b.AllocateMemory(4)
-			b.Push(value.S32(f.loopCount))
-			b.Store(f.loopCountPtr)
-			b.JumpLabel(uint32(0x1))
-			return nil
-		}))
-		out.NotifyPreLoop(ctx)
-
-	} else if cmd == f.loopEndCmd {
-		log.D(ctx, "Loop: last frame is %v id %v, cmd is %v.", f.frameNum, id, cmd)
-		st := GetState(out.State())
-		sb := st.newStateBuilder(ctx, newTransformerOutput(out))
-		defer sb.ta.Dispose()
-		out.MutateAndWrite(ctx, id, cmd)
-
-		// Notify this is the end part of the loop to next transformer
-		out.NotifyPostLoop(ctx)
-		if err := f.resetResource(ctx, sb); err != nil {
-			log.E(ctx, "Failed to reset changed resources %v.", err)
+
+	} else if f.loopTerminated == false { // We're not before or at the start of the loop: thus, are we inside the loop or just at the end of it?
+
+		// This is the end of the loop. We have a lot of deferred things to do.
+		if api.CmdID.Real(cmdId) >= f.loopEndIdx {
+
+			if lastObservedCommand == api.CmdNoID {
+				log.F(ctx, true, "FrameLoop: Somehow, the FrameLoop ended before it began. Did an earlier transform delete the whole loop? Were your loop indexes realistic?")
+			}
+
+			if len(f.capturedLoopCmdIds) != len(f.capturedLoopCmds) {
+				log.F(ctx, true, "FrameLoop: Control flow error: Somehow, the number of captured commands and commandIds are not equal.")
+			}
+
+			f.loopTerminated = true
+			log.D(ctx, "FrameLoop: last frame is %v cmdId %v, cmd is %v.", f.frameNum, cmdId, cmd)
+
+			// Do start loop stuff.
```

Comment:
not sure how this is done inside the `if api.CmdID.Real(cmdId) >= f.loopEndIdx {` branch

Reply:
This is the "stop queuing things, and start actually running commands"case. So when we are at the end, then we deal with queued commands.

Reply:
then the queueing action probably need to be done outside of the `Transfer` method. I think when you reached the `loopEndIdx`, the mutation from later transfer already done?

Reply:
@AWoloszyn any follow up on Robin's comment? I think this is correct, but to be honest I have not followed the whole Transformer system though end to end.

Reply:
@RenfengLiu The transform from later transforms is only done when you call "out.MutateAndWrite", meaning I think we are safe here. This just  queues up commands and doesn't call out.MutateAndWrite.